go/contacts/resolution_notice.go

```diff
@@ -0,0 +1,145 @@
+)
+
+type ContactResolution struct {
+	Description  string
+	ResolvedUser keybase1.User
+}
+
+const ContactResolutionEncryptionVersion = 1
+
+type ContactResolutionForEncryptrion struct {
+	ContactResolution
+	EncryptionVersion int
+}
+
+type EncryptedContactResolution struct {
+	Blob          []byte
+	PukGeneration keybase1.PerUserKeyGeneration
+}
+
+var errWrongContactEncryptionScheme = errors.New(
+	"wrong encryption version for encrypted contact blob")
+
+func getKeyFn(mctx libkb.MetaContext,
+	pukGen *keybase1.PerUserKeyGeneration) (keyFn func(ctx context.
+	Context) ([32]byte, error), gen keybase1.PerUserKeyGeneration,
+	errOuter error) {
+	pukring, err := mctx.G().GetPerUserKeyring(mctx.Ctx())
+	if err != nil {
+		return nil, gen, err
+	}
+	err = pukring.Sync(mctx)
+	if err != nil {
+		return nil, gen, err
+	}
+	currentGen := pukring.CurrentGeneration()
+	if pukGen == nil {
+		pukGen = &currentGen
+	}
+	keyFn = func(ctx context.Context) (res [32]byte, err error) {
+		keypair, err := pukring.GetEncryptionKeyByGeneration(mctx, *pukGen)
+
+		// Derive symmetric key from PUK
+		skey, err := keypair.SecretSymmetricKey(libkb.EncryptionReasonContactsResolvedServer)
+		if err != nil {
+			return res, err
+		}
+
+		copy(res[:], skey[:])
+		return res, nil
+	}
+	return keyFn, *pukGen, nil
+}
+
+func encryptContactBlob(mctx libkb.MetaContext, res ContactResolution) (string,
+	error) {
+	keyFn, gen, err := getKeyFn(mctx, nil)
+	if err != nil {
+		return "", err
+	}
+	boxedRes := ContactResolutionForEncryptrion{res, ContactResolutionEncryptionVersion}
+	encrypted, err := encrypteddb.EncodeBox(mctx.Ctx(), boxedRes, keyFn)
+	if err != nil {
+		return "", err
+	}
+	result := EncryptedContactResolution{
+		Blob:          encrypted,
+		PukGeneration: gen,
+	}
+	messagePacked, err := codec.MessagePackEncode(result)
+	if err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(messagePacked), nil
+}
+
+func DecryptContactBlob(mctx libkb.MetaContext,
+	contactResBlob string) (res ContactResolution, err error) {
+	messagePacked, err := base64.StdEncoding.DecodeString(contactResBlob)
+	if err != nil {
+		return res, err
+	}
+	var unpacked EncryptedContactResolution
+	err = codec.MessagePackDecode(messagePacked, &unpacked)
+	if err != nil {
+		return res, err
+	}
+	keyFn, _, err := getKeyFn(mctx, &unpacked.PukGeneration)
+	if err != nil {
+		return res, err
+	}
+	var boxedRes ContactResolutionForEncryptrion
+	err = encrypteddb.DecodeBox(mctx.Ctx(), unpacked.Blob,
+		keyFn, &boxedRes)
+	if err != nil {
+		return res, err
+	}
+	if boxedRes.EncryptionVersion != ContactResolutionEncryptionVersion {
+		return res, errWrongContactEncryptionScheme
+	}
+	return boxedRes.ContactResolution, nil
+}
+
+// TODO: actually call this
+func SendContactResolutionToServer(mctx libkb.MetaContext,
+	resolutions []ContactResolution) error {
+
+	type resolvedArg struct {
+		ResolvedContactBlobBase64 string `json:"b"`
```

Comment:
this is fine, but in general you don't have to minimize the json names unless you're really trying to save on space